control/Director.php

```diff
@@ -385,10 +385,15 @@ protected static function handleRequest(SS_HTTPRequest $request, Session $sessio
 						$result = $controllerObj->handleRequest($request, $model);
 					} catch(SS_HTTPResponse_Exception $responseException) {
 						$result = $responseException->getResponse();
-						// If catching controller exception, retroactively add cache-breaking for errors
+					}
+					// Ensure cache headers are added
+					if ($result instanceof SS_HTTPResponse) {
 						HTTP::add_cache_headers($result);
+						return $result;
+					}
+					if(!is_object($result)) {
+						return $result;
```

Comment:
Shouldn't we call `HTTP::add_cache_headers()` without an argument in this case?

Reply:
That would likely break a lot behaviour with too-early http header output.

Reply:
I *believe* headers are only flushed once actual output is started, so shouldn't be too big a deal